First, we import everything we need and connect to the Reddit API

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelBinarizer

from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LassoCV
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import ElasticNetCV
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
import warnings
warnings.filterwarnings('ignore')

import os
from dotenv import load_dotenv
load_dotenv()

CLIENT_ID = os.getenv('CLIENT_ID')
CLIENT_SECRET=os.getenv('CLIENT_SECRET')
APP_NAME=os.getenv('APP_NAME')
REDDIT_USERNAME=os.getenv('REDDIT_USERNAME')
REDDIT_PASSWORD=os.getenv('REDDIT_PASSWORD')

import praw
import pandas as pd
import datetime as dt

reddit = praw.Reddit(client_id=CLIENT_ID, client_secret=CLIENT_SECRET, user_agent=APP_NAME, username=REDDIT_USERNAME, password=REDDIT_PASSWORD)

print(reddit.user.me())

%matplotlib inline

mattcat26


Next, we are going to read in the bad comment data given to us by a moderator of the 'neutralnews' subreddit.

In [34]:
import gzip
import shutil
with gzip.open('./neutralnews-2020-09-27.json.gz', 'rb') as f_in:
    with open('neutralnews-2020-09-27.json', 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)

In [35]:
comment_df = pd.read_json('neutralnews-2020-09-27.json', orient='records')

In [36]:
new_comment_df = comment_df[["action", "content", "author", "details", "submissionId", "id"]]
bad_comment_data = new_comment_df.rename(columns={"id" : "commentId"})
bad_comment_data


action                                            content  \
0      removecomment  How is this"neutral news"? Literally, it's rep...   
1      removecomment            Just wondering if you have any updates.   
2      removecomment                                                  .   
3      removecomment                                     I went to park   
4      removecomment  Hi, u/SFepicure\n\n[I think you may enjoy this...   
...              ...                                                ...   
5997   removecomment  Oh I thought we were posting arbitrary facts t...   
5998   removecomment  No.  The senate is not wasting time to fill th...   
5999  approvecomment  My point was not addressed, regarding providin...   
6000   removecomment  Voting has begun in some states but the debate...   
6001  approvecomment  Voting has begun in some states but the debate...   

                                author  \
0     b756df5867ce042f3a07f3037e5eeeb9   
1     58843f7430c71f72208766a295eaae5e   
2     f9dcef98ec140a8f44d5691d34081408   
3     aab8641b0dc26f59e9c8fea95f470138   
4     e48fe0135108ee9985e260caf57cc6e0   
...                                ...   
5997  38e2216dc3dc2cb196cbf4bab0d83541   
5998  c4f22a6532a4ca274c5027e4bc3d3d99   
5999  062512f6041e1c40256c31f31a1138fc   
6000  c696933b8cbd4b625077d6b577bac0fe   
6001  c696933b8cbd4b625077d6b577bac0fe   

                                     details submissionId  \
0     Rule 5: top-level comment has no links       cgd7ut   
1               Low effort top-level comment       cim6kf   
2     Rule 5: top-level comment has no links       cfzkky   
3               Low effort top-level comment       asrdao   
4                                     remove       b47a1d   
...                                      ...          ...   
5997                                  remove       j0ijwn   
5998                                  remove       j0ijwn   
5999                                  unspam       j04eqv   
6000                                  remove       j0ijwn   
6001                                  unspam       j0ijwn   

                     commentId  
0     5dea4ee94af43200093a2f4c  
1     5dea4ee94af43200093a2f4d  
2     5dea4ee94af43200093a2f4e  
3     5dea4ee94af43200093a2f4f  
4     5dea4ee94af43200093a2f50  
...                        ...  
5997  5f70e968dbe1ef0009d976e7  
5998  5f70e968dbe1ef0009d976e9  
5999  5f70ee18dbe1ef0009d97717  
6000  5f70f3f4dbe1ef0009d9775e  
6001  5f70f3f4dbe1ef0009d97760  

[6002 rows x 6 columns]

In [37]:
new_arr = bad_comment_data['submissionId'].to_list()
new_arr = np.array(new_arr)
submissions = np.unique(new_arr)
submissions = submissions.tolist()
submissions

['4o2o29',
 '4op948',
 '4sef35',
 '5x0k84',
 '64zsim',
 '68qqfz',
 '6907e3',
 '6mmw91',
 '6n161i',
 '6oftmx',
 '6xapy0',
 '7307wv',
 '76z4xh',
 '7ht8tj',
 '7hyg56',
 '7z5dyt',
 '804o3f',
 '862sn6',
 '8d0t2m',
 '8e7lrz',
 '8go559',
 '8hz10r',
 '8uby76',
 '8x2571',
 '8zp0uh',
 '91gwrp',
 '94q8j7',
 '96nrmf',
 '99fccg',
 '9c88xn',
 '9f2i6b',
 '9fj3m8',
 '9h0mo6',
 '9h82yb',
 '9ht83g',
 '9iqed1',
 '9j3gvd',
 '9jpdh7',
 '9kq8xy',
 '9ood1n',
 '9rbpt8',
 '9s3gd8',
 '9smhsl',
 '9xyyt0',
 '9zonfn',
 'a4tn3i',
 'a67z1j',
 'aab5gq',
 'ab7m8y',
 'ac7d56',
 'acz2f9',
 'ahxiba',
 'ailn92',
 'aj96gr',
 'ajvsq6',
 'alzqd0',
 'an9m5u',
 'andfa8',
 'apss2q',
 'asacv6',
 'asrdao',
 'auz29y',
 'axnhrj',
 'b47a1d',
 'b7e8o4',
 'b8cd1q',
 'b9d62h',
 'bj3tgb',
 'bjngf5',
 'bka66t',
 'blxd09',
 'bpi7u6',
 'busrz0',
 'bv8dl8',
 'bvk72a',
 'bxtj3y',
 'byiwll',
 'c109ai',
 'c3swuk',
 'c75b2i',
 'c7nqp7',
 'cess35',
 'cfa7tc',
 'cfa99a',
 'cfc5qa',
 'cfzkky',
 'cgd7ut',
 'cgks97',
 'cgkspq',
 'cgktg9',
 'cgkueu',

In [38]:
topics_dict = {'title':[], 'score':[], 'id':[], 'url':[], 'comms_num': [], 'created':[], 'body':[]}

for list_submission in submissions:
    submission = reddit.submission(id=list_submission)
    topics_dict['title'].append(submission.title)
    topics_dict['score'].append(submission.score)
    topics_dict['id'].append(submission.id)
    topics_dict['url'].append(submission.url)
    topics_dict['comms_num'].append(submission.num_comments)
    topics_dict['created'].append(submission.created)
    topics_dict['body'].append(submission.selftext)
topics_data = pd.DataFrame(topics_dict)

Now, we are going to parse in good comment data by scraping the neutral news subreddit.

In [39]:
comments_dict = {"action": [], "content": [], "author": [], "details": [], "submissionId": [], "commentId": []}

In [40]:
subreddit = reddit.subreddit('neutralnews')

for list_submission in submissions:
    submission = reddit.submission(id=list_submission)
    #print(submission.title, submission.id)
    submission.comments.replace_more(limit=100)
    for comment in submission.comments:
        #print(top_level_comment.body)
        comments_dict["action"].append(np.nan)
        comments_dict["content"].append(comment.body)
        comments_dict["author"].append(comment.author)
        comments_dict["details"].append(np.nan)
        comments_dict["submissionId"].append(submission.id)
        comments_dict["commentId"].append(comment.id)

good_comment_data = pd.DataFrame(comments_dict)

In [41]:
good_comment_data = good_comment_data[good_comment_data['author'] != 'AutoModerator']

In [42]:
good_comment_data = good_comment_data.reset_index()

In [43]:
good_comment_data = good_comment_data.drop(['index'], axis=1)
good_comment_data

action                                            content  \
0        NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1        NaN  Hi, thanks for making this sub. A couple quest...   
2        NaN  I'd like to say thanks to our "focus group" pa...   
3        NaN                                          [removed]   
4        NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...      ...                                                ...   
2606     NaN  r/NeutralNews is a **curated space**, but desp...   
2607     NaN  __I'm a bot. Here are The Factual credibility ...   
2608     NaN                                          [removed]   
2609     NaN                                          [removed]   
2610     NaN  Judiciary hearing starts oct 12th.  Fill the s...   

               author  details submissionId commentId  
0             zaphnod      NaN       4o2o29   d498hvk  
1           niugnep24      NaN       4o2o29   d49b7t2  
2            nosecohn      NaN       4o2o29   d498cxc  
3                None      NaN       4o2o29   d4997am  
4          bloodguard      NaN       4o2o29   d49d04t  
...               ...      ...          ...       ...  
2606  NeutralverseBot      NaN       j0ijwn   g6rh4no  
2607    TheFactualBot      NaN       j0ijwn   g6rhg7f  
2608             None      NaN       j0ijwn   g6ud0cs  
2609             None      NaN       j0ijwn   g6ukt21  
2610        met021345      NaN       j0ijwn   g6rn19i  

[2611 rows x 6 columns]

We can now combine the good comment data and bad comment data into one dataframe.

In [44]:
from datetime import datetime
now = datetime.now()
current_time = now.strftime('%H:%M:%S')
print('current time = ', current_time)

current time =  21:41:29


After combining the good comment data, run sam's code on the whole comment data and then run the upvote predictor code on the features

In [47]:
good_comment_data.to_csv("good_comment_data_119.csv", index=False)
bad_comment_data.to_csv("bad_comment_data_119.csv", index=False)

In [6]:
from newspaper import Article
from newspaper import Config

user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36'
config = Config()
config.browser_user_agent = user_agent

import spacy
from spacy import displacy
from collections import Counter
#!python -m spacy download en_core_web_lg
#!pip install spacy-wordnet

import en_core_web_lg
nlp = en_core_web_lg.load()

from spacy_wordnet.wordnet_annotator import WordnetAnnotator
nlp.add_pipe(WordnetAnnotator(nlp.lang), after='tagger')


STOP_WORDS = spacy.lang.en.stop_words.STOP_WORDS

✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_lg')


In [52]:
frames = [good_comment_data, bad_comment_data]
comment_data = pd.concat(frames, ignore_index=True)
comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
3                NaN                                          [removed]   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
3                                 None     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  
0                      d498hvk  
1                      d49b7t2  
2                      d498cxc  
3                      d4997am  
4                      d49d04t  
...                        ...  
8608  5f70e968dbe1ef0009d976e7  
8609  5f70e968dbe1ef0009d976e9  
8610  5f70ee18dbe1ef0009d97717  
8611  5f70f3f4dbe1ef0009d9775e  
8612  5f70f3f4dbe1ef0009d97760  

[8613 rows x 6 columns]

In [7]:
good_comment_data = pd.read_csv('good_comment_data_119.csv')
bad_comment_data = pd.read_csv('bad_comment_data_119.csv')
frames = [good_comment_data, bad_comment_data]
comment_data = pd.concat(frames, ignore_index=True)
comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
3                NaN                                          [removed]   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
3                                  NaN     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  
0                      d498hvk  
1                      d49b7t2  
2                      d498cxc  
3                      d4997am  
4                      d49d04t  
...                        ...  
8608  5f70e968dbe1ef0009d976e7  
8609  5f70e968dbe1ef0009d976e9  
8610  5f70ee18dbe1ef0009d97717  
8611  5f70f3f4dbe1ef0009d9775e  
8612  5f70f3f4dbe1ef0009d97760  

[8613 rows x 6 columns]

In [93]:
import nltk
nltk.download('wordnet')

from datetime import datetime
now = datetime.now()
current_time = now.strftime('%H:%M:%S')
print('current time = ', current_time)

def clean_articles(topics_data, comment_data, text_list):
    for url in topics_data['url']:
        try:
            article = Article(url, language='en', fetch_images=False, config = config)
            article.download()
            article.parse()
            art_text = article.text
            art_doc = nlp(art_text.lower())
            text_list.append(art_doc)
        #if there is an exception, remove comments that go with this article
        except:
            text_list.append("error")
            subId = topics_data[topics_data['url'] == url]['id']
            subId = subId.to_numpy()
            subId = subId[0]
            comment_data = comment_data[comment_data['submissionId'] != subId]
            continue
    topics_data['text'] = text_list

clean_articles(topics_data, comment_data, [])
# url = 'https://www.washingtonpost.com/world/national-security/sessions-spoke-twice-with-russian-ambassador-during-trumps-presidential-campaign-justice-officials-say/2017/03/01/77205eda-feac-11e6-99b4-9e613afeb09f_story.html?hpid=hp_rhp-top-table-main_no-name%3Ahomepage%2Fstory&utm_term=.f2a14c329123'
# subId = topics_data[topics_data['url'] == url]['id']
# subId = subId.to_numpy()
# subId = subId[0]
# subId

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\mattc\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
current time =  23:16:59


In [94]:
from datetime import datetime
now = datetime.now()
current_time = now.strftime('%H:%M:%S')
print('current time = ', current_time)

current time =  23:40:41


In [96]:
topics_data.to_csv("topics_data_119.csv", index=False)

In [8]:
topics_data = pd.read_csv('topics_data_119.csv')
topics_data

title  score      id  \
0                        [META] Welcome to NeutralNews    225  4o2o29   
1    Wall Street has been rocked by an $8 billion h...    183  4op948   
2    How a $2 Roadside Drug Test Sends Innocent Peo...     91  4sef35   
3    Sessions spoke twice with Russian ambassador d...    668  5x0k84   
4    The Russia story just keeps getting worse for ...     90  64zsim   
..                                                 ...    ...     ...   
531  Trump falsely claimed an incident where an ele...      9  j078ii   
532  ‘It’s like every red flag’: Trump-ordered HHS ...      5  j099m5   
533  Fact-Checking Falsehoods on Mail-In Voting - V...     15  j09b0v   
534  How a Pledge to Dismantle the Minneapolis Poli...      3  j0h85d   
535  It's Official: Trump Nominates Judge Amy Coney...     95  j0ijwn   

                                                   url  comms_num  \
0    https://www.reddit.com/r/neutralnews/comments/...         66   
1    http://www.businessinsider.com/visium-asset-ma...         10   
2    http://www.nytimes.com/2016/07/10/magazine/how...         11   
3    https://www.washingtonpost.com/world/national-...         84   
4    http://www.cnn.com/2017/04/12/politics/trump-c...         31   
..                                                 ...        ...   
531  https://www.businessinsider.com/trump-falsely-...         39   
532  https://www.politico.com/news/2020/09/25/trump...          4   
533  https://www.nytimes.com/article/fact-checking-...          3   
534  https://www.nytimes.com/2020/09/26/us/politics...          3   
535  https://reason.com/2020/09/26/its-official-tru...         42   

          created                                               body  \
0    1.465955e+09  The goal of /r/NeutralNews is to provide a pla...   
1    1.466297e+09                                                NaN   
2    1.468315e+09                                                NaN   
3    1.488450e+09                                                NaN   
4    1.492048e+09                                                NaN   
..            ...                                                ...   
531  1.601160e+09                                                NaN   
532  1.601167e+09                                                NaN   
533  1.601167e+09                                                NaN   
534  1.601195e+09                                          [deleted]   
535  1.601200e+09                                                NaN   

                                                  text  
0    the goal of r/neutralnews is to provide a plac...  
1    jake gottlieb, the founder of visium. reuters/...  
2    field tests provide quick answers. but if thos...  
3                                                error  
4    washington (cnn) two stories dealing with russ...  
..                                                 ...  
531  president donald trump falsely claimed that an...  
532  the $15 million contract, which has not been p...  
533  if you are among the tens of millions of ameri...  
534  minneapolis — over three months ago, a majorit...  
535  president donald trump on saturday announced t...  

[536 rows x 8 columns]

In [18]:
#nlp = spacy.load('en_core_web_lg')
def getSimWordScore(comment_data, topics_data, simWordScore):
    for index, comment in comment_data.iterrows():
        subID = comment['submissionId'] #get submission id from comment table
        post = topics_data[topics_data['id'] == subID]
        art_doc = post['text']
        
        art_doc = art_doc.to_numpy()
        art_doc = art_doc[0]
        art_doc = nlp(str(art_doc))
        #art_doc = nlp(art_doc)
        art_items = [x.text for x in art_doc.ents]
        #get tokens
        art_tokens = []
        for (item, count) in Counter(art_items).most_common(5):
            token = nlp(item)[0]
            art_tokens += [token]
        #get comment content and ner
        comment_text = comment['content']
        comment_text = str(comment_text)
        doc = nlp(str(comment_text).lower())

        items = [x.text for x in doc.ents]

        #initialize list of scores
        score = 0
        #for each token, get a score
        for (item, count) in Counter(items).most_common(5):

            #get token
            token = nlp(item)#[0]

            wordScores = []

            #for each article item
            for art_word in art_tokens:

                #add similarity score to list of scores
                wordScores += [art_word.similarity(token)]
            #get average score
            if len(wordScores) != 0:
                score += sum(wordScores)/len(wordScores)
            else:
                score = 0
        simWordScore.append(score)
    return simWordScore

wordScoreList = getSimWordScore(comment_data, topics_data, [])
comment_data['WordScore'] = wordScoreList

In [19]:
from datetime import datetime
now = datetime.now()
current_time = now.strftime('%H:%M:%S')
print('current time = ', current_time)

current time =  11:16:40


In [21]:
comment_data.to_csv("comment_data_1110.csv", index=False)

In [29]:
def getSimWholeScore(comment_data, topics_data, simWholeScore):
    for index, comment in comment_data.iterrows():
        subID = comment['submissionId']
        #get article text
        post = topics_data[topics_data['id'] == subID]
        art_doc = str(post['text'])
        art_doc = nlp(art_doc)
        #get comment content and ner
        comment_text = comment['content']
        comment_text = str(comment).lower()
        doc = nlp(comment_text)
        #compare to get a score
        simWholeScore.append(art_doc.similarity(doc))
    return simWholeScore

wholeScoreList = getSimWholeScore(comment_data, topics_data, [])
comment_data['WholeScore'] = wholeScoreList

In [30]:
from datetime import datetime
now = datetime.now()
current_time = now.strftime('%H:%M:%S')
print('current time = ', current_time)

current time =  12:15:46


In [32]:
comment_data.to_csv("comment_data_1110.csv", index=False)

In [33]:
comment_data.to_csv('full_comment_data_with_features119.csv', index=False)

# Final cleaning of data (hopefully)

In [3]:
#Wanted to compare full data with original combo of data
good_comment_data = pd.read_csv('good_comment_data_119.csv')
bad_comment_data = pd.read_csv('bad_comment_data_119.csv')
frames = [good_comment_data, bad_comment_data]
comment_data = pd.concat(frames, ignore_index=True)
comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
3                NaN                                          [removed]   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
3                                  NaN     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  
0                      d498hvk  
1                      d49b7t2  
2                      d498cxc  
3                      d4997am  
4                      d49d04t  
...                        ...  
8608  5f70e968dbe1ef0009d976e7  
8609  5f70e968dbe1ef0009d976e9  
8610  5f70ee18dbe1ef0009d97717  
8611  5f70f3f4dbe1ef0009d9775e  
8612  5f70f3f4dbe1ef0009d97760  

[8613 rows x 6 columns]

In [18]:
final_comment_data = pd.read_csv('full_comment_data_with_features119.csv')
final_comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
3                NaN                                          [removed]   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
3                                  NaN     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  WordScore  WholeScore  
0                      d498hvk   0.055391    0.869869  
1                      d49b7t2   0.000000    0.896084  
2                      d498cxc   0.371254    0.877248  
3                      d4997am   0.000000    0.810516  
4                      d49d04t   0.000000    0.897533  
...                        ...        ...         ...  
8608  5f70e968dbe1ef0009d976e7   0.000000    0.838691  
8609  5f70e968dbe1ef0009d976e9   0.459704    0.840920  
8610  5f70ee18dbe1ef0009d97717   1.281814    0.863294  
8611  5f70f3f4dbe1ef0009d9775e   0.000000    0.820983  
8612  5f70f3f4dbe1ef0009d97760   0.000000    0.830435  

[8613 rows x 8 columns]

### Final data is same size as original combo (8613 rows), so have some bad data yet

In [19]:
topics_data = pd.read_csv('topics_data_119.csv')
topics_data

title  score      id  \
0                        [META] Welcome to NeutralNews    225  4o2o29   
1    Wall Street has been rocked by an $8 billion h...    183  4op948   
2    How a $2 Roadside Drug Test Sends Innocent Peo...     91  4sef35   
3    Sessions spoke twice with Russian ambassador d...    668  5x0k84   
4    The Russia story just keeps getting worse for ...     90  64zsim   
..                                                 ...    ...     ...   
531  Trump falsely claimed an incident where an ele...      9  j078ii   
532  ‘It’s like every red flag’: Trump-ordered HHS ...      5  j099m5   
533  Fact-Checking Falsehoods on Mail-In Voting - V...     15  j09b0v   
534  How a Pledge to Dismantle the Minneapolis Poli...      3  j0h85d   
535  It's Official: Trump Nominates Judge Amy Coney...     95  j0ijwn   

                                                   url  comms_num  \
0    https://www.reddit.com/r/neutralnews/comments/...         66   
1    http://www.businessinsider.com/visium-asset-ma...         10   
2    http://www.nytimes.com/2016/07/10/magazine/how...         11   
3    https://www.washingtonpost.com/world/national-...         84   
4    http://www.cnn.com/2017/04/12/politics/trump-c...         31   
..                                                 ...        ...   
531  https://www.businessinsider.com/trump-falsely-...         39   
532  https://www.politico.com/news/2020/09/25/trump...          4   
533  https://www.nytimes.com/article/fact-checking-...          3   
534  https://www.nytimes.com/2020/09/26/us/politics...          3   
535  https://reason.com/2020/09/26/its-official-tru...         42   

          created                                               body  \
0    1.465955e+09  The goal of /r/NeutralNews is to provide a pla...   
1    1.466297e+09                                                NaN   
2    1.468315e+09                                                NaN   
3    1.488450e+09                                                NaN   
4    1.492048e+09                                                NaN   
..            ...                                                ...   
531  1.601160e+09                                                NaN   
532  1.601167e+09                                                NaN   
533  1.601167e+09                                                NaN   
534  1.601195e+09                                          [deleted]   
535  1.601200e+09                                                NaN   

                                                  text  
0    the goal of r/neutralnews is to provide a plac...  
1    jake gottlieb, the founder of visium. reuters/...  
2    field tests provide quick answers. but if thos...  
3                                                error  
4    washington (cnn) two stories dealing with russ...  
..                                                 ...  
531  president donald trump falsely claimed that an...  
532  the $15 million contract, which has not been p...  
533  if you are among the tens of millions of ameri...  
534  minneapolis — over three months ago, a majorit...  
535  president donald trump on saturday announced t...  

[536 rows x 8 columns]

In [20]:
removed_topics = topics_data[topics_data['text'] == "error"]
removed_topics

title  score      id  \
3    Sessions spoke twice with Russian ambassador d...    668  5x0k84   
34   ‘Kavanaugh will be on the U.S. Supreme Court’:...    317  9ht83g   
39   Google really is trying to build a censored Ch...    458  9ood1n   
40   Judge orders Ga. officials to stop tossing abs...     76  9rbpt8   
74   Justice Department does not comply with court ...    317  bvk72a   
93   Update on NN in This Post: [META] r/NeutralPol...     68  f8says   
123  In unusual deal, U.S. Treasury to acquire 30 p...    214  hjimwh   
151  Evictions are likely to skyrocket this summer ...     36  hmor9m   
165  Can Trump delay the release of his taxes to Ma...     11  ho63fl   
171  Robert Mueller: Roger Stone remains a convicte...    369  hpjk4b   
197  The Latest: WHO posts 2nd daily record for new...    201  htol34   
247  Trump ordered federal forces to quell Portland...    554  i16cum   
253  St. Louis prosecutor to detective on McCloskey...      1  i21mky   
261  How the Trump campaign came to court QAnon, th...     58  i2f8b3   
267  Colorado police apologize over viral video of ...    129  i3k5uu   
303  ‘This is no longer a debate’: Florida sheriff ...    344  i8nfh1   
311  Postal Service warns 46 states their voters co...    431  i9smf6   
315  Trump, Biden tout experience in police chiefs’...     32  ia8tf4   
347  Tennessee adopts new law that could strip some...    399  iewpfo   
353  New York attorney general files legal action a...    328  ifwln1   
368  Video shows activists in Kenosha arrested by o...    254  ihz2hg   
380  Trump leans into his ‘180,000 deaths is a stat...    159  ik07zh   
383  Biden calls Trump ‘a toxic presence’ who is en...    252  ikbrh9   
389  Pelosi Takes Heat Over Visit to California Hai...     10  il0iua   
414  At least 7 police officers suspended in Roches...    332  im36cp   
445  ICE flew detainees to Virginia so the planes c...     85  iqus5t   
462  FBI says no raid was conducted at home of cons...     25  istm8e   
470  Pro-Trump group pays teens in secretive online...     11  itmffo   
481  Newly revealed USPS documents show an agency s...     28  iuvl82   
492  Trump’s appointees sought to censor what gover...    383  ivy8qp   
503  CDC reverses itself and says guidelines it pos...     35  ix6rxb   
515  Trump won't commit to peaceful transfer of pow...     10  iyoo0x   
525  Trump administration rescinded Courage Award f...    296  izrbbb   

                                                   url  comms_num  \
3    https://www.washingtonpost.com/world/national-...         84   
34   https://www.washingtonpost.com/politics/2018/0...        173   
39   https://www.washingtonpost.com/technology/2018...         73   
40   https://www.washingtonpost.com/politics/judge-...          7   
74   https://www.washingtonpost.com/politics/justic...         13   
93   /r/NeutralPolitics/comments/f3fyvr/meta_rneutr...         10   
123  https://www.washingtonpost.com/business/2020/0...         25   
151  https://www.washingtonpost.com/business/2020/0...         17   
165  https://www.washingtonpost.com/politics/courts...          7   
171  https://www.washingtonpost.com/opinions/2020/0...         48   
197  https://apnews.com/3f595fc278431a2e6c5b165d8b4...         14   
247  https://www.washingtonpost.com/nation/2020/07/...        151   
253  https://www.ksdk.com/article/news/crime/i-sugg...          5   
261  https://www.washingtonpost.com/politics/how-th...         12   
267  https://www.washingtonpost.com/nation/2020/08/...         13   
303  https://www.washingtonpost.com/nation/2020/08/...        134   
311  https://www.washingtonpost.com/local/md-politi...         47   
315  https://www.washingtonpost.com/crime-law/2020/...          3   
347  https://www.washingtonpost.com/politics/tennes...        137   
353  https://www.washingtonpost.com/politics/new-yo...         24   
368  https://www.washingtonpost.com/nation/2020/08/...         91   
380  https://www.washingtonpost.com/politics/202

In [21]:
removed_comments = final_comment_data[final_comment_data['submissionId'].isin(removed_topics['id'])]

In [22]:
removed_comments

action                                            content  \
33               NaN                                          [removed]   
34               NaN  Here is [Sessions' response](https://twitter.c...   
35               NaN  Question: how does this not come out during th...   
36               NaN                                          [removed]   
37               NaN                                          [removed]   
...              ...                                                ...   
8443  approvecomment        That wasnt the question the reporter asked.   
8444  approvecomment  But in this thread we have specifically called...   
8478   removecomment                                          Good bot!   
8479   removecomment  Over exaggerating fear mongering title. He won...   
8480   removecomment  Sounds par for the course for this administrat...   

                                author      details submissionId  \
33                                 NaN          NaN       5x0k84   
34                IDrink_n_IKnowThings          NaN       5x0k84   
35                       such-a-mensch          NaN       5x0k84   
36                                 NaN          NaN       5x0k84   
37                                 NaN          NaN       5x0k84   
...                                ...          ...          ...   
8443  c4f22a6532a4ca274c5027e4bc3d3d99  confirm_ham       iyoo0x   
8444  38e2216dc3dc2cb196cbf4bab0d83541       unspam       iyoo0x   
8478  0a0c22e1c4803b776901400bd71cdceb        Memes       izrbbb   
8479  a1397b06b9fffbe73d9cb8c77bc1e1ae       remove       iyoo0x   
8480  9b08adf8b2671b5410ead0483b50a0cf       remove       izrbbb   

                     commentId  WordScore  WholeScore  
33                     deegcr4        0.0    0.819557  
34                     deegxs1        0.0    0.842452  
35                     deerirr        0.0    0.800308  
36                     deedont        0.0    0.818399  
37                     deenab8        0.0    0.816503  
...                        ...        ...         ...  
8443  5f6d1107dbe1ef0009d95b3e        0.0    0.783706  
8444  5f6d1b93dbe1ef0009d95b85        0.0    0.780085  
8478  5f6e9d20dbe1ef0009d967fd        0.0    0.818553  
8479  5f6ece57dbe1ef0009d968e4        0.0    0.748502  
8480  5f6ed7b7dbe1ef0009d96917        0.0    0.786327  

[705 rows x 8 columns]

In [23]:
final_comment_data = final_comment_data[~final_comment_data['submissionId'].isin(removed_topics['id'])]
final_comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
3                NaN                                          [removed]   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
3                                  NaN     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  WordScore  WholeScore  
0                      d498hvk   0.055391    0.869869  
1                      d49b7t2   0.000000    0.896084  
2                      d498cxc   0.371254    0.877248  
3                      d4997am   0.000000    0.810516  
4                      d49d04t   0.000000    0.897533  
...                        ...        ...         ...  
8608  5f70e968dbe1ef0009d976e7   0.000000    0.838691  
8609  5f70e968dbe1ef0009d976e9   0.459704    0.840920  
8610  5f70ee18dbe1ef0009d97717   1.281814    0.863294  
8611  5f70f3f4dbe1ef0009d9775e   0.000000    0.820983  
8612  5f70f3f4dbe1ef0009d97760   0.000000    0.830435  

[7908 rows x 8 columns]

In [24]:
final_comment_data = final_comment_data[final_comment_data['content'] != "[removed]"]
final_comment_data = final_comment_data[final_comment_data['content'] != "[deleted]"]
final_comment_data

action                                            content  \
0                NaN  Just wanted to say "thanks" publicly.  I'm ver...   
1                NaN  Hi, thanks for making this sub. A couple quest...   
2                NaN  I'd like to say thanks to our "focus group" pa...   
4                NaN  Can we get "Moderator Activity Log" for /r/Neu...   
5                NaN  Will political and election news be deleted si...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
0                              zaphnod     NaN       4o2o29   
1                            niugnep24     NaN       4o2o29   
2                             nosecohn     NaN       4o2o29   
4                           bloodguard     NaN       4o2o29   
5                                  NaN     NaN       4o2o29   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  WordScore  WholeScore  
0                      d498hvk   0.055391    0.869869  
1                      d49b7t2   0.000000    0.896084  
2                      d498cxc   0.371254    0.877248  
4                      d49d04t   0.000000    0.897533  
5                      d497qee   0.000000    0.854950  
...                        ...        ...         ...  
8608  5f70e968dbe1ef0009d976e7   0.000000    0.838691  
8609  5f70e968dbe1ef0009d976e9   0.459704    0.840920  
8610  5f70ee18dbe1ef0009d97717   1.281814    0.863294  
8611  5f70f3f4dbe1ef0009d9775e   0.000000    0.820983  
8612  5f70f3f4dbe1ef0009d97760   0.000000    0.830435  

[7257 rows x 8 columns]

In [25]:
final_comment_data = final_comment_data[final_comment_data['submissionId'] != '4o2o29']
final_comment_data

action                                            content  \
28               NaN  So what are the implications here? Does it onl...   
30               NaN  Sadly this isn't new. Police officers use many...   
31               NaN  What's disturbing about this is that our gover...   
32               NaN  What I find really concerning is the horrible ...   
38               NaN  This subject might have legs but this article ...   
...              ...                                                ...   
8608   removecomment  Oh I thought we were posting arbitrary facts t...   
8609   removecomment  No.  The senate is not wasting time to fill th...   
8610  approvecomment  My point was not addressed, regarding providin...   
8611   removecomment  Voting has begun in some states but the debate...   
8612  approvecomment  Voting has begun in some states but the debate...   

                                author details submissionId  \
28                          Cody_Fox23     NaN       4op948   
30                         DrFrenchman     NaN       4sef35   
31                              bbakks     NaN       4sef35   
32                        poliscijunki     NaN       4sef35   
38                           interweb1     NaN       64zsim   
...                                ...     ...          ...   
8608  38e2216dc3dc2cb196cbf4bab0d83541  remove       j0ijwn   
8609  c4f22a6532a4ca274c5027e4bc3d3d99  remove       j0ijwn   
8610  062512f6041e1c40256c31f31a1138fc  unspam       j04eqv   
8611  c696933b8cbd4b625077d6b577bac0fe  remove       j0ijwn   
8612  c696933b8cbd4b625077d6b577bac0fe  unspam       j0ijwn   

                     commentId  WordScore  WholeScore  
28                     d4eictg   0.000000    0.849655  
30                     d58ts90   0.000000    0.900283  
31                     d58y081  -0.038865    0.869078  
32                     d5919n8   0.000000    0.898426  
38                     dg6l969   0.000000    0.850127  
...                        ...        ...         ...  
8608  5f70e968dbe1ef0009d976e7   0.000000    0.838691  
8609  5f70e968dbe1ef0009d976e9   0.459704    0.840920  
8610  5f70ee18dbe1ef0009d97717   1.281814    0.863294  
8611  5f70f3f4dbe1ef0009d9775e   0.000000    0.820983  
8612  5f70f3f4dbe1ef0009d97760   0.000000    0.830435  

[7231 rows x 8 columns]

In [26]:
final_comment_data.to_csv('full_comment_data_with_features1110.csv', index=False)

### Wanted to see amount of bad comments

In [35]:
good_comment_length = final_comment_data[pd.isna(final_comment_data['action'])]
good_comment_length

action                                            content  \
28      NaN  So what are the implications here? Does it onl...   
30      NaN  Sadly this isn't new. Police officers use many...   
31      NaN  What's disturbing about this is that our gover...   
32      NaN  What I find really concerning is the horrible ...   
38      NaN  This subject might have legs but this article ...   
...     ...                                                ...   
2604    NaN  __I'm a bot. Here are The Factual credibility ...   
2605    NaN  They had a real chance to do something new and...   
2606    NaN  r/NeutralNews is a **curated space**, but desp...   
2607    NaN  __I'm a bot. Here are The Factual credibility ...   
2610    NaN  Judiciary hearing starts oct 12th.  Fill the s...   

               author details submissionId commentId  WordScore  WholeScore  
28         Cody_Fox23     NaN       4op948   d4eictg   0.000000    0.849655  
30        DrFrenchman     NaN       4sef35   d58ts90   0.000000    0.900283  
31             bbakks     NaN       4sef35   d58y081  -0.038865    0.869078  
32       poliscijunki     NaN       4sef35   d5919n8   0.000000    0.898426  
38          interweb1     NaN       64zsim   dg6l969   0.000000    0.850127  
...               ...     ...          ...       ...        ...         ...  
2604    TheFactualBot     NaN       j0h85d   g6r4cxh   1.512839    0.818242  
2605        911roofer     NaN       j0h85d   g6r7c8r   0.000000    0.855738  
2606  NeutralverseBot     NaN       j0ijwn   g6rh4no   0.080495    0.791657  
2607    TheFactualBot     NaN       j0ijwn   g6rhg7f   0.811500    0.815048  
2610        met021345     NaN       j0ijwn   g6rn19i   0.154348    0.855474  

[1785 rows x 8 columns]

In [ ]:
#1785 good comments